python/tchannel/tchannel.py

```diff
@@ -0,0 +1,135 @@
+from __future__ import absolute_import
+
+import logging
+import os
+import sys
+import socket
+
+import tornado.ioloop
+import tornado.iostream
+
+from .tornado.connection import TornadoConnection
+from .exceptions import InvalidMessageException
+from .messages import CallRequestMessage
+
+
+log = logging.getLogger('tchannel')
+
+
+class TChannel(object):
+    """Manages inbound and outbound connections to various hosts."""
+
+    def __init__(self):
+        self.peers = {}
+        self.process_name = "%s[%s]" % (sys.argv[0], os.getpid())
+
+        # TODO: self.server_socket = self.make_in_connection()
+
+    @tornado.gen.coroutine
+    def add_peer(self, hostport):
+        if hostport in self.peers:
+            return
+
+        peer = yield self.make_out_connection(hostport)
+        self.peers[hostport] = peer
+        raise tornado.gen.Return(peer)
+
+    def remove_peer(self, hostport):
+        # TODO: Connection cleanup
+        self.peers.pop(hostport)
+
+    @tornado.gen.coroutine
+    def get_peer(self, hostport):
+        if hostport in self.peers:
+            peer = self.peers[hostport]
+        else:
+            peer = yield self.add_peer(hostport)
+        raise tornado.gen.Return(peer)
+
+    @tornado.gen.coroutine
+    def make_out_connection(self, hostport):
+        host, port = hostport.split(":")
+
+        sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+        stream = tornado.iostream.IOStream(sock)
+
+        log.debug("connecting to hostport %s", hostport)
+
+        yield stream.connect((host, int(port)))
+
+        connection = TornadoConnection(stream)
+
+        log.debug("initiating handshake with %s", sock.getsockname())
+
+        yield connection.initiate_handshake(headers={
+            'host_port': '%s:%s' % sock.getsockname(),
+            'process_name': self.process_name,
+        })
+
+        log.debug("awaiting handshake reply")
+
+        yield connection.await_handshake_reply()
+
+        log.debug("completed handshake")
+
+        raise tornado.gen.Return(connection)
+
+    @tornado.gen.coroutine
+    def make_in_connection(self):
+        raise NotImplementedError
+
+    def listen(self):
+        # TODO: listen to inbound handshakes
+        raise NotImplementedError
+
+    def respond(self):
+        # TODO: handle inbound requests
+        raise NotImplementedError
+
+    @tornado.gen.coroutine
+    def request(self, hostport):
+
+        peer_connection = yield self.get_peer(hostport)
+
+        class TChannelClientOp(object):
```

Comment:
What's the advantage of this being a nested class instead of an external class which accepts the `peer_connection` in its constructor?